File: services/go-api/internal/handlers/subscription.go
package handlers

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
	"github.com/MargoRSq/infatium-mono/services/go-api/internal/clients"
	"github.com/MargoRSq/infatium-mono/services/go-api/internal/middleware"
	"github.com/MargoRSq/infatium-mono/services/go-api/repository"
	"github.com/rs/zerolog/log"
)

type SubscriptionHandler struct {
	subscriptionRepo *repository.SubscriptionRepository
	usersFeedRepo    *repository.UsersFeedRepository
	ruStoreClient    *clients.RuStoreClient
}

func NewSubscriptionHandler(
	subscriptionRepo *repository.SubscriptionRepository,
	usersFeedRepo *repository.UsersFeedRepository,
	ruStoreClient *clients.RuStoreClient,
) *SubscriptionHandler {
	return &SubscriptionHandler{
		subscriptionRepo: subscriptionRepo,
		usersFeedRepo:    usersFeedRepo,
		ruStoreClient:    ruStoreClient,
	}
}

func (h *SubscriptionHandler) Routes() chi.Router {
	r := chi.NewRouter()

	r.Get("/current", h.GetCurrentSubscription)
	r.Get("/limits", h.GetLimits)
	r.Post("/validate", h.ValidateSubscription)
	r.Post("/manual", h.CreateManualSubscription)

	return r
}

type CurrentSubscriptionResponse struct {
	HasActiveSubscription bool         `json:"has_active_subscription"`
	Plan                  PlanResponse `json:"plan"`
	ActiveFeedsCount      int          `json:"active_feeds_count"`
	SubscriptionID        *uuid.UUID   `json:"subscription_id,omitempty"`
	Platform              *string      `json:"platform,omitempty"`
	StartDate             *string      `json:"start_date,omitempty"`
	ExpiryDate            *string      `json:"expiry_date,omitempty"`
	IsAutoRenewing        bool         `json:"is_auto_renewing"`
	Status                *string      `json:"status,omitempty"`
}

type PlanResponse struct {
	ID                  uuid.UUID `json:"id"`
	PlanType            string    `json:"plan_type"`
	FeedsLimit          int       `json:"feeds_limit"`
	SourcesPerFeedLimit int       `json:"sources_per_feed_limit"`
	PriceAmountMicros   *int64    `json:"price_amount_micros,omitempty"`
	IsActive            bool      `json:"is_active"`
	CreatedAt           string    `json:"created_at"`
}

func (h *SubscriptionHandler) GetCurrentSubscription(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	activeFeedsCount, err := h.usersFeedRepo.CountUserFeeds(r.Context(), userID)
	if err != nil {
		log.Error().Err(err).Msg("Failed to count user feeds")
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	sub, err := h.subscriptionRepo.GetCurrentSubscription(r.Context(), userID)
	if err != nil {
		log.Error().Err(err).Msg("Failed to get current subscription")
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	if sub == nil {
		freePlan, err := h.subscriptionRepo.GetFreePlan(r.Context())
		if err != nil || freePlan == nil {
			writeJSON(w, http.StatusOK, CurrentSubscriptionResponse{
				HasActiveSubscription: false,
				Plan: PlanResponse{
					ID:                  uuid.Nil,
					PlanType:            "FREE",
					FeedsLimit:          3,
					SourcesPerFeedLimit: 5,
					IsActive:            true,
					CreatedAt:           time.Now().Format(time.RFC3339),
				},
				ActiveFeedsCount: activeFeedsCount,
				IsAutoRenewing:   false,
			})
			return
		}

		var priceAmountMicros *int64
		if freePlan.PriceAmountMicros > 0 {
			priceAmountMicros = &freePlan.PriceAmountMicros
		}

		writeJSON(w, http.StatusOK, CurrentSubscriptionResponse{
			HasActiveSubscription: false,
			Plan: PlanResponse{
				ID:                  freePlan.ID,
				PlanType:            freePlan.PlanType,
				FeedsLimit:          freePlan.FeedsLimit,
				SourcesPerFeedLimit: freePlan.SourcesPerFeedLimit,
				PriceAmountMicros:   priceAmountMicros,
				IsActive:            freePlan.IsActive,
				CreatedAt:           time.Now().Format(time.RFC3339),
			},
			ActiveFeedsCount: activeFeedsCount,
			IsAutoRenewing:   false,
		})
		return
	}

	plan, err := h.subscriptionRepo.GetPlanByID(r.Context(), sub.SubscriptionPlanID)
	if err != nil || plan == nil {
		log.Error().Err(err).Msg("Failed to get subscription plan")
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	var expiryDate *string
	if sub.ExpiryDate != nil {
		exp := sub.ExpiryDate.Format(time.RFC3339)
		expiryDate = &exp
	}

	startDate := sub.StartDate.Format(time.RFC3339)

	var priceAmountMicros *int64
	if plan.PriceAmountMicros > 0 {
		priceAmountMicros = &plan.PriceAmountMicros
	}

	writeJSON(w, http.StatusOK, CurrentSubscriptionResponse{
		HasActiveSubscription: true,
		Plan: PlanResponse{
			ID:                  plan.ID,
			PlanType:            plan.PlanType,
			FeedsLimit:          plan.FeedsLimit,
			SourcesPerFeedLimit: plan.SourcesPerFeedLimit,
			PriceAmountMicros:   priceAmountMicros,
			IsActive:            plan.IsActive,
			CreatedAt:           sub.CreatedAt.Format(time.RFC3339),
		},
		ActiveFeedsCount: activeFeedsCount,
		SubscriptionID:   &sub.ID,
		Platform:         &sub.Platform,
		StartDate:        &startDate,
		ExpiryDate:       expiryDate,
		IsAutoRenewing:   sub.IsAutoRenewing,
		Status:           &sub.Status,
	})
}

type LimitsResponse struct {
	PlanType              string  `json:"plan_type"`
	FeedsLimit            int     `json:"feeds_limit"`
	SourcesPerFeedLimit   int     `json:"sources_per_feed_limit"`
	HasActiveSubscription bool    `json:"has_active_subscription"`
	ExpiryDate            *string `json:"expiry_date,omitempty"`
}

func (h *SubscriptionHandler) GetLimits(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	sub, err := h.subscriptionRepo.GetCurrentSubscription(r.Context(), userID)
	if err != nil {
		log.Error().Err(err).Msg("Failed to get subscription")
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	if sub != nil {
		plan, err := h.subscriptionRepo.GetPlanByID(r.Context(), sub.SubscriptionPlanID)
		if err != nil || plan == nil {
			log.Error().Err(err).Msg("Failed to get plan")
			http.Error(w, "internal error", http.StatusInternalServerError)
			return
		}

		var expiryDate *string
		if sub.ExpiryDate != nil {
			exp := sub.ExpiryDate.Format(time.RFC3339)
			expiryDate = &exp
		}

		writeJSON(w, http.StatusOK, LimitsResponse{
			PlanType:              plan.PlanType,
			FeedsLimit:            plan.FeedsLimit,
			SourcesPerFeedLimit:   plan.SourcesPerFeedLimit,
			HasActiveSubscription: true,
			ExpiryDate:            expiryDate,
		})
		return
	}

	freePlan, _ := h.subscriptionRepo.GetFreePlan(r.Context())
	if freePlan != nil {
		writeJSON(w, http.StatusOK, LimitsResponse{
			PlanType:              freePlan.PlanType,
			FeedsLimit:            freePlan.FeedsLimit,
			SourcesPerFeedLimit:   freePlan.SourcesPerFeedLimit,
			HasActiveSubscription: false,
			ExpiryDate:            nil,
		})
		return
	}

	writeJSON(w, http.StatusOK, LimitsResponse{
		PlanType:              "FREE",
		FeedsLimit:            3,
		SourcesPerFeedLimit:   5,
		HasActiveSubscription: false,
		ExpiryDate:            nil,
	})
}

type CreateManualSubscriptionRequest struct {
	UserID   uuid.UUID `json:"user_id"`
	Months   int       `json:"months"`
	PlanType string    `json:"plan_type"`
}

type ManualSubscriptionResponse struct {
	Success        bool      `json:"success"`
	SubscriptionID uuid.UUID `json:"subscription_id"`
	PlanType       string    `json:"plan_type"`
	Status         string    `json:"status"`
	ExpiryDate     string    `json:"expiry_date"`
	Message        string    `json:"message"`
}

func (h *SubscriptionHandler) CreateManualSubscription(w http.ResponseWriter, r *http.Request) {
	var req CreateManualSubscriptionRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "invalid request body", http.StatusBadRequest)
		return
	}

	if req.UserID == uuid.Nil {
		http.Error(w, "user_id is required", http.StatusBadRequest)
		return
	}

	if req.Months <= 0 {
		req.Months = 1
	}

	if req.PlanType == "" {
		req.PlanType = "PRO"
	}

	plan, err := h.subscriptionRepo.GetPlanByType(r.Context(), req.PlanType)
	if err != nil || plan == nil {
		http.Error(w, "plan not found", http.StatusNotFound)
		return
	}

	if err := h.subscriptionRepo.DeactivateUserSubscriptions(r.Context(), req.UserID); err != nil {
		log.Error().Err(err).Msg("Failed to deactivate existing subscriptions")
	}

	expiresAt := time.Now().AddDate(0, req.Months, 0)

	sub, err := h.subscriptionRepo.Create(r.Context(), repository.CreateSubscriptionParams{
		UserID:    req.UserID,
		PlanID:    plan.ID,
		Status:    "ACTIVE",
		Platform:  "MANUAL",
		ExpiresAt: &expiresAt,
	})
	if err != nil {
		log.Error().Err(err).Msg("Failed to create subscription")
		http.Error(w, "internal error", http.StatusInternalServerError)
		return
	}

	writeJSON(w, http.StatusCreated, ManualSubscriptionResponse{
		Success:        true,
		SubscriptionID: sub.ID,
		PlanType:       plan.PlanType,
		Status:         sub.Status,
		ExpiryDate:     expiresAt.Format(time.RFC3339),
		Message:        fmt.Sprintf("Subscription created for %d months", req.Months),
	})
}

type ValidateSubscriptionRequest struct {
	PackageName    string `json:"package_name"`
	SubscriptionID string `json:"subscription_id"`
	PurchaseToken  string `json:"purchase_token"`
}

type ValidateSubscriptionResponse struct {
	Success        bool       `json:"success"`
	SubscriptionID *uuid.UUID `json:"subscription_id,omitempty"`
	PlanType       *string    `json:"plan_type,omitempty"`
	Status         *string    `json:"status,omitempty"`
	ExpiryDate     *string    `json:"expiry_date,omitempty"`
	Message        string     `json:"message"`
}

func (h *SubscriptionHandler) ValidateSubscription(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	var req ValidateSubscriptionRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "invalid request body", http.StatusBadRequest)
		return
	}

	if req.PackageName == "" || req.SubscriptionID == "" || req.PurchaseToken == "" {
		writeJSON(w, http.StatusBadRequest, ValidateSubscriptionResponse{
			Success: false,
			Message: "package_name, subscription_id, and purchase_token are required",
		})
		return
	}

	log.Info().
		Str("user_id", userID.String()).
		Str("subscription_id", req.SubscriptionID).
		Msg("Validating subscription")

	if h.ruStoreClient == nil {
		writeJSON(w, http.StatusServiceUnavailable, ValidateSubscriptionResponse{
			Success: false,
			Message: "RuStore integration not configured",
		})
		return
	}

	ruStoreResp, err := h.ruStoreClient.ValidateSubscription(r.Context(), req.PackageName, req.SubscriptionID, req.PurchaseToken)
	if err != nil {
		log.Error().Err(err).Msg("RuStore validation failed")
		writeJSON(w, http.StatusBadRequest, ValidateSubscriptionResponse{
			Success: false,
			Message: fmt.Sprintf("Subscription validation failed: %s", err.Error()),
		})
		return
	}

	if ruStoreResp.PaymentState != 1 {
		log.Warn().
			Int("payment_state", ruStoreResp.PaymentState).
			Msg("Subscription payment state is not active")
		writeJSON(w, http.StatusBadRequest, ValidateSubscriptionResponse{
			Success: false,
			Message: "Subscription payment is not active",
		})
		return
	}

	existingSub, _ := h.subscriptionRepo.GetByPlatformSubscriptionID(r.Context(), req.PurchaseToken)
	if existingSub != nil {
		expiryDateStr := existingSub.ExpiryDate.Format(time.RFC3339)
		status := existingSub.Status
		plan, _ := h.subscriptionRepo.GetPlanByID(r.Context(), existingSub.SubscriptionPlanID)
		planType := "PRO"
		if plan != nil {
			planType = plan.PlanType
		}
		writeJSON(w, http.StatusOK, ValidateSubscriptionResponse{
			Success:        true,
			SubscriptionID: &existingSub.ID,
			PlanType:       &planType,
			Status:         &status,
			ExpiryDate:     &expiryDateStr,
			Message:        "Subscription already exists",
		})
		return
	}

	proPlan, err := h.subscriptionRepo.GetPlanByType(r.Context(), "PRO")
	if err != nil || proPlan == nil {
		log.Error().Err(err).Msg("Failed to get PRO plan")
		writeJSON(w, http.StatusInternalServerError, ValidateSubscriptionResponse{
			Success: false,
			Message: "Failed to get subscription plan",
		})
		return
	}

	if err := h.subscriptionRepo.DeactivateUserSubscriptions(r.Context(), userID); err != nil {
		log.Error().Err(err).Msg("Failed to deactivate existing subscriptions")
	}

	expiryTime := time.Unix(0, ruStoreResp.ExpiryTimeMillis*int64(time.Millisecond))
	externalID := req.PurchaseToken

	sub, err := h.subscriptionRepo.Create(r.Context(), repository.CreateSubscriptionParams{
		UserID:     userID,
		PlanID:     proPlan.ID,
		Status:     "ACTIVE",
		Platform:   "RUSTORE",
		ExternalID: &externalID,
		ExpiresAt:  &expiryTime,
	})
	if err != nil {
		log.Error().Err(err).Msg("Failed to create subscription")
		writeJSON(w, http.StatusInternalServerError, ValidateSubscriptionResponse{
			Success: false,
			Message: "Failed to create subscription",
		})
		return
	}

	expiryDateStr := expiryTime.Format(time.RFC3339)
	status := "ACTIVE"
	planType := proPlan.PlanType

	writeJSON(w, http.StatusOK, ValidateSubscriptionResponse{
		Success:        true,
		SubscriptionID: &sub.ID,
		PlanType:       &planType,
		Status:         &status,
		ExpiryDate:     &expiryDateStr,
		Message:        "Subscription validated successfully",
	})
}

File: services/shared-python/shared/models/subscription.py
"""Pydantic models for subscription operations."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from shared.enums import (
    SubscriptionPlanType,
    SubscriptionPlatform,
    SubscriptionStatus,
)


class SubscriptionPlanResponse(BaseModel):
    """Subscription plan information."""

    id: UUID = Field(examples=["550e8400-e29b-41d4-a716-446655440001"])
    plan_type: SubscriptionPlanType = Field(examples=["PRO", "FREE"])
    feeds_limit: int = Field(examples=[10, 3, 50])
    sources_per_feed_limit: int = Field(examples=[20, 5, 100])
    price_amount_micros: int | None = Field(
        default=None, examples=[4990000, 9990000, None]
    )
    is_active: bool = Field(examples=[True, False])
    created_at: datetime = Field(examples=["2024-01-01T00:00:00Z"])

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "id": "550e8400-e29b-41d4-a716-446655440001",
                    "plan_type": "PRO",
                    "feeds_limit": 10,
                    "sources_per_feed_limit": 20,
                    "price_amount_micros": 4990000,
                    "is_active": True,
                    "created_at": "2024-01-01T00:00:00Z",
                }
            ]
        }
    )


class ValidateSubscriptionRequest(BaseModel):
    """Request to validate RuStore subscription."""

    package_name: str = Field(
        description="Application package name (e.g., com.example.app)",
        examples=["ru.infatium.makefeed", "com.example.app"],
    )
    subscription_id: str = Field(
        description="Product/subscription ID from RuStore Console",
        examples=["makefeed_pro_monthly", "premium_yearly"],
    )
    purchase_token: str = Field(
        description="Subscription token from RuStore SDK (format: invoiceId.userId)",
        examples=["INV123456789.USR987654321"],
    )

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "package_name": "ru.infatium.makefeed",
                    "subscription_id": "makefeed_pro_monthly",
                    "purchase_token": "INV123456789.USR987654321",
                }
            ]
        }
    )


class ValidateSubscriptionResponse(BaseModel):
    """Response from subscription validation."""

    success: bool = Field(examples=[True, False])
    subscription_id: UUID | None = Field(
        default=None, examples=["550e8400-e29b-41d4-a716-446655440002", None]
    )
    plan_type: SubscriptionPlanType | None = Field(
        default=None, examples=["PRO", "FREE", None]
    )
    status: SubscriptionStatus | None = Field(
        default=None, examples=["ACTIVE", "EXPIRED", None]
    )
    expiry_date: datetime | None = Field(
        default=None, examples=["2024-02-15T10:30:00Z", None]
    )
    message: str = Field(
        examples=["Subscription validated successfully", "Invalid purchase token"]
    )

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "success": True,
                    "subscription_id": "550e8400-e29b-41d4-a716-446655440002",
                    "plan_type": "PRO",
                    "status": "ACTIVE",
                    "expiry_date": "2024-02-15T10:30:00Z",
                    "message": "Subscription validated successfully",
                }
            ]
        }
    )


class CurrentSubscriptionResponse(BaseModel):
    """Current subscription information for user."""

    has_active_subscription: bool = Field(examples=[True, False])
    plan: SubscriptionPlanResponse = Field(
        examples=[
            {
                "id": "550e8400-e29b-41d4-a716-446655440001",
                "plan_type": "PRO",
                "feeds_limit": 10,
                "sources_per_feed_limit": 20,
                "price_amount_micros": 4990000,
                "is_active": True,
                "created_at": "2024-01-01T00:00:00Z",
            }
        ]
    )
    active_feeds_count: int = Field(
        description="Current number of active feeds for the user",
        examples=[3, 0, 10],
    )
    subscription_id: UUID | None = Field(
        default=None, examples=["550e8400-e29b-41d4-a716-446655440002", None]
    )
    platform: SubscriptionPlatform | None = Field(
        default=None, examples=["RUSTORE", "APPSTORE", None]
    )
    start_date: datetime | None = Field(
        default=None, examples=["2024-01-15T10:30:00Z", None]
    )
    expiry_date: datetime | None = Field(
        default=None, examples=["2024-02-15T10:30:00Z", None]
    )
    is_auto_renewing: bool = Field(default=False, examples=[True, False])
    status: SubscriptionStatus | None = Field(
        default=None, examples=["ACTIVE", "EXPIRED", None]
    )

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "has_active_subscription": True,
                    "plan": {
                        "id": "550e8400-e29b-41d4-a716-446655440001",
                        "plan_type": "PRO",
                        "feeds_limit": 10,
                        "sources_per_feed_limit": 20,
                        "price_amount_micros": 4990000,
                        "is_active": True,
                        "created_at": "2024-01-01T00:00:00Z",
                    },
                    "active_feeds_count": 3,
                    "subscription_id": "550e8400-e29b-41d4-a716-446655440002",
                    "platform": "RUSTORE",
                    "start_date": "2024-01-15T10:30:00Z",
                    "expiry_date": "2024-02-15T10:30:00Z",
                    "is_auto_renewing": True,
                    "status": "ACTIVE",
                }
            ]
        }
    )


class SubscriptionWithLimitsResponse(BaseModel):
    """Subscription information with usage limits."""

    plan_type: SubscriptionPlanType = Field(examples=["PRO", "FREE"])
    feeds_limit: int = Field(examples=[10, 3, 50])
    sources_per_feed_limit: int = Field(examples=[20, 5, 100])
    has_active_subscription: bool = Field(examples=[True, False])
    expiry_date: datetime | None = Field(
        default=None, examples=["2024-02-15T10:30:00Z", None]
    )

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "plan_type": "PRO",
                    "feeds_limit": 10,
                    "sources_per_feed_limit": 20,
                    "has_active_subscription": True,
                    "expiry_date": "2024-02-15T10:30:00Z",
                }
            ]
        }
    )


class ManualSubscriptionRequest(BaseModel):
    """Request for manual subscription creation (admin)."""

    user_id: UUID = Field(examples=["550e8400-e29b-41d4-a716-446655440003"])
    months: int = Field(
        default=1, ge=1, description="Duration in months", examples=[1, 3, 12]
    )
    plan_type: SubscriptionPlanType = Field(
        default=SubscriptionPlanType.PRO,
        description="Subscription plan type",
        examples=["PRO"],
    )

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "user_id": "550e8400-e29b-41d4-a716-446655440003",
                    "months": 3,
                    "plan_type": "PRO",
                }
            ]
        }
    )


class ManualSubscriptionResponse(BaseModel):
    """Response for manual subscription creation."""

    success: bool = Field(examples=[True, False])
    subscription_id: UUID = Field(examples=["550e8400-e29b-41d4-a716-446655440004"])
    plan_type: SubscriptionPlanType = Field(examples=["PRO", "FREE"])
    status: SubscriptionStatus = Field(examples=["ACTIVE", "PENDING"])
    expiry_date: datetime = Field(examples=["2024-04-15T10:30:00Z"])
    message: str = Field(
        examples=[
            "Subscription created for 3 months",
            "Subscription extended successfully",
        ]
    )

    model_config = ConfigDict(
        json_schema_extra={
            "examples": [
                {
                    "success": True,
                    "subscription_id": "550e8400-e29b-41d4-a716-446655440004",
                    "plan_type": "PRO",
                    "status": "ACTIVE",
                    "expiry_date": "2024-04-15T10:30:00Z",
                    "message": "Subscription created for 3 months",
                }
            ]
        }
    )
